optimizely/event/dispatcher.go

```diff
@@ -54,10 +55,72 @@ func (*HTTPEventDispatcher) DispatchEvent(event LogEvent, callback func(bool)) {
 		if resp.StatusCode == 204 {
 			success = true
 		} else {
-			fmt.Printf("http.Post invalid response %d", resp.StatusCode)
+			dispatcherLogger.Error(fmt.Sprintf("http.Post invalid response %d", resp.StatusCode), nil)
 			success = false
 		}
 	}
 	callback(success)
+}
+
+// QueueEventDispatcher is a queued version of the event dispatcher that queues, returns success, and dispatches events in the background
+type QueueEventDispatcher struct {
+	eventQueue     Queue
+	eventFlushLock sync.Mutex
+	dispatcher     *HTTPEventDispatcher
+}
+
+// DispatchEvent queues event with callback and calls flush in a go routine.
+func (ed *QueueEventDispatcher) DispatchEvent(event LogEvent, callback func(bool)) {
+
+	ed.eventQueue.Add(event)
+	callback(true)
+	go func() {
+		ed.flushEvents()
+	}()
+}
+
+// flush the events
+func (ed *QueueEventDispatcher) flushEvents() {
+
+	ed.eventFlushLock.Lock()
+
+	defer func(){
+		ed.eventFlushLock.Unlock()
+	}()
+
+	for ed.eventQueue.Size() > 0 {
+		items := ed.eventQueue.Get(1)
+		if len(items) == 0 {
```

Comment:
why do we need to wait 5 seconds here ? why len(items) == 0 when ed.eventQueue.Size() > 0 ? are they not related ?

Reply:
i am just checking the array size after the call for safety.

Reply:
just a sanity check.  not sure we need the sleep in there.

Reply:
ok, if this is a sanity check, what should we do if len(items) == 0 ? that means something went wrong, should we report it ? stop the process? I think repeater pattern would be more elegant here, unless you want to try forever by design. I don't think sleep is needed here

Reply:
I took out the sleep and just continue with the expectation that size will return 0 or more and I can try again